File: JavaScript/webserver.js
/*
 * A simple Web server implementation in JavaScript (on Node.js)
 *
 * Usage:
 *   node webserver.js
 *
 * Author: Kota Abe
 */

// "http"モジュールを使うともっと簡単にWebサーバを実装できるが，
// このプログラムでは低レベルの処理を見せるために"net"モジュールを使っている
const net = require("net");

const SERVER_PORT = 8888;

function main() {
    const server = net.createServer(); // サーバの作成
    // JavaScriptでは基本的にイベントベースで処理を記述する
    // - クライアントからのコネクションが確立すると"connection"というイベントが発生する．
    // - そのときに conn => { handleClient(conn); } というアロー関数が実行されるように設定する．
    //    connは確立したコネクション(net.Socketのオブジェクト)
    // https://nodejs.org/api/net.html#net_event_connection
    server.on("connection", conn => {
        handleClient(conn);
    });
    server.listen(SERVER_PORT); // コネクションの待ち受けを開始する
    console.log(`open http://localhost:${SERVER_PORT}/ with your browser!`);
    // ここでmain関数は終わるが，裏でコネクションの確立を待っている
}

/**
 * クライアントとHTTP/1.0で通信する．
 * @param {net.Socket} conn
 */
 function handleClient(conn) {
    console.log(`Connection from ${conn.remoteAddress}:${conn.remotePort} has been established!`)
    let buf = "";  // 読み込んだHTTPヘッダを覚えておくバッファ
    // クライアントからデータが届いた場合の処理を設定する
    // (設定するだけなので，この処理はデータが届く前に終了する)
    conn.on("data", data => {
        // この部分はデータが届いたときに実行される
        console.log("Received: " + data);
        buf = buf + data; // データは複数回に別れて届く可能性があるので，受け取ったデータを連結する
        const i = buf.indexOf("\r\n\r\n"); // 空行があるかどうかを判定
        if (i >= 0) { // 空行がある場合
            const head = buf.slice(0, i); // 先頭から空行までを取り出す   
            const headers = head.split("\r\n"); // 行単位に分割する
            // 先頭行を " " で分割する．分割代入構文を使って配列の要素を3つの変数に代入
            let [ method, path, http_version ] = headers[0].split(" ");
            console.log(`method=${method}, path=${path}, http_version=${http_version}`);
            switch (method) {
                case "GET":
                    handleGet(conn, path);
                    break;
                default:
                    console.log(`unsupported method: ${method}`);
                    conn.write("HTTP/1.0 501 Not Implemented\r\n");
                    break;
            }
            conn.destroy(); // コネクションをクローズ
        }
    });
    // コネクションがクローズされた場合の処理を設定する
    conn.on("end", () => {
        console.log("connection closed");
        conn.destroy();
    });
}

/**
 * GET要求を処理する
 */
 function handleGet(conn, path) {
    var s
    if (path === "/") {
        // backquote(``)では文字列中の改行は\nになる
        s = `HTTP/1.0 200 OK
Content-Type: text/html

<!DOCTYPE html>
<html>
<head><title>Sample</title></head>
<body>This server is implemented in JavaScript!</body>
</html>
`
    } else {
        s = `HTTP/1.0 404 Not Found
Content-Type: text/html

<!DOCTYPE html>
<html>
<head><title>404 Not Found</title></head>
<body>${path} is not found</body>
</html>
`
    }
  	// 改行を LF (\n) から CR+LF (\r\n) に置き換える
    replaced = s.replace("\n", "\r\n")
    conn.write(replaced)
}

main();

File: Elixir/webserver.exs
defmodule WebServer do
  @moduledoc """
  A simple Web server implementation in Elixir

  Usage: elixir ./webserver.exs

  Author: Kota Abe
  """
  def main(_args \\ []) do
    server_port = 8888
    {:ok, socket} = :gen_tcp.listen(server_port,
      [:binary, packet: :line, active: false, reuseaddr: true])
    IO.puts("open http://localhost:#{server_port}/ with your browser!")
    loop_acceptor(socket)
  end

  @doc """
  クライアントからのコネクションを待ち受ける
  """
  def loop_acceptor(socket) do
    # コネクションを待ち受ける
    {:ok, client} = :gen_tcp.accept(socket)
    # 相手のIPアドレスとポート番号を取得
    {:ok, {address, port}} = :inet.peername(client)
    IO.puts("Connection from #{:inet.ntoa(address)}:#{port} has been established!")
    # handle_client関数を軽量プロセスとして実行 (concurrentサーバ)
    spawn(__MODULE__, :handle_client, [client])
    # iterativeサーバの場合は spawn の代わりに以下のように関数呼び出しする
    # handle_client(client)
    loop_acceptor(socket) # 繰り返し(再帰)
  end

  @doc """
  クライアントとHTTP/1.0で通信
  """
  def handle_client(client) do
    # read_headerでヘッダを読み取る
    case read_header(client, []) do
      {:ok, headers} ->
        req_line = Enum.at(headers, 0, "")  # 1行目(リクエスト行)を取り出す
        case String.split(req_line) do      # リクエスト行を空白で分割
          ["GET", path, _http_version] -> handle_get(client, path)
          [method, _path, _http_version] ->
            IO.puts("unsupported method: #{method}")
            :gen_tcp.send(client, "HTTP/1.0 501 Not Implemented\r\n")
          _ ->
            IO.puts("wrong request: #{req_line}")
        end
      _ -> nil
    end
    :gen_tcp.close(client)
    :ok
  end

  @doc """
  クライアントからリクエスト行とHTTPヘッダを読み込む
  """
  def read_header(client, headers) do
    case :gen_tcp.recv(client, 0) do        # 1行読み込む
      {:ok, line} ->                        # 成功した場合
        line = String.trim_trailing(line)   # 末尾の改行を削除
        IO.puts("Received: #{line}")
        if line == "" do
          {:ok, headers}                    # 空行なら終了
        else
          # 空行でなければ読み込みを続ける(再帰)
          read_header(client, headers ++ [line])
        end
      {:error, reason} ->                   # 失敗した場合
        IO.puts("gen_tcp.recv failed (#{reason})")
        {:error, reason}
    end
  end

  @doc """
  GETリクエストを処理
  """
  def handle_get(client, path) do
    s = case path do
      "/" -> """
HTTP/1.0 200 OK
Content-Type: text/html

<!DOCTYPE html>
<html>
  <head><title>Sample</title></head>
  <body>This server is implemented in Elixir!</body>
</html>
"""
      _ -> """
HTTP/1.0 404 Not Found
Content-Type: text/html

<!DOCTYPE html>
<html>
  <head><title>404 Not Found</title></head>
  <body>#{path} is not found</body>
</html>
"""
    end
    # 改行を LF (\n) から CR+LF (\r\n) に置き換える
    replaced = String.replace(s, "\n", "\r\n", global: true)
    :gen_tcp.send(client, replaced)
  end
end

WebServer.main()
